src/qtils/bytes.hpp

```diff
@@ -11,17 +11,146 @@
 #include <span>
 #include <vector>
 
+#include <qtils/assert.hpp>
+
 namespace qtils {
   using Bytes = std::vector<uint8_t>;
   template <size_t N>
   using BytesN = std::array<uint8_t, N>;
   using BytesIn = std::span<const uint8_t>;
   using BytesOut = std::span<uint8_t>;
 
+  template <size_t N>
+  using ByteArray = std::array<uint8_t, N>;
+
+  using ByteSpan = std::span<const unsigned char>;
+
+  template <size_t N>
+  using FixedByteSpan = std::span<const unsigned char, N>;
+
+  template <size_t N>
+  using FixedByteSpanMut = std::span<unsigned char, N>;
+
+  using ByteSpanMut = std::span<unsigned char>;
+
   template <typename T>
   concept AsBytes = requires(const T &t) { BytesIn{t}; };
 
   inline Bytes asVec(BytesIn v) {
     return {v.begin(), v.end()};
   }
+
+  /**
+   * Fixed-capacity vector meant to be allocated on stack
+   * Used to avoid heap allocations for small vectors with known capacity
+   * @tparam N - maximum capacity
+   */
+  template <size_t N>
+  struct FixedByteVector {
+    BytesN<N> data;
+    size_t size;
+
+    FixedByteVector() : data{}, size{} {}
+
+    FixedByteVector(ByteSpan span) : size{span.size()} {
+      QTILS_ASSERT_LESS_EQ(span.size(), N);
+      std::copy_n(span.begin(), N, data.begin());
+    }
+
+    ByteSpan span() const {
+      return ByteSpan{data.data(), size};
+    }
+  };
+
+  inline std::string_view to_string_view(ByteSpan span) {
+    return std::string_view{
+        reinterpret_cast<const char *>(span.data()), span.size()};
+  }
+
+  inline std::string hex(ByteSpan span) {
+    std::string s(span.size() * 2, '\0');
+    for (size_t i = 0; i < span.size(); i++) {
+      std::to_chars(s.data() + 2 * i, s.data() + 2 * i + 1, span[i] >> 4, 16);
+      std::to_chars(
+          s.data() + 2 * i + 1, s.data() + 2 * i + 2, span[i] & 0xF, 16);
+    }
+    return s;
+  }
+
+  template <size_t N>
+  BytesN<N> array_from_span(ByteSpan span) {
```

Comment:
Move `array_from_span` and `array_from_span_fill` to separate header ("array_from_span.hpp"?)

Reply:
Why?